lib/core/sync_job_runner.rb

```diff
@@ -94,19 +90,31 @@ def do_sync!
 
         Utility::Logger.debug("#{existing_ids.size} documents are present in index #{@index_name}.")
 
-        post_processing_engine = Core::Filtering::PostProcessEngine.new(@job.filtering)
-        Utility::Logger.info('Yielding documents...')
-        @connector_instance.yield_documents do |document|
-          document = add_ingest_metadata(document)
-          post_process_result = post_processing_engine.process(document)
-          if post_process_result.is_include?
-            @sink.ingest(document)
-            incoming_ids << document['id']
+        # We want to validate advanced filtering rules even if basic rules are disabled
+        if @connector_settings.any_filtering_feature_enabled?
+          Utility::Logger.info("Checking active filtering for sync job #{@job_id} for connector #{@connector_id}.")
+          validate_filtering(@job.filtering)
+          Utility::Logger.debug("Active filtering for sync job #{@job_id} for connector #{@connector_id} is valid.")
+        end
+
+        if @connector_settings.filtering_rule_feature_enabled?
+          post_processing_engine = Core::Filtering::PostProcessEngine.new(@job.filtering)
+
+          Utility::Logger.info('Yielding documents with basic rule filtering...')
+
+          yield_docs do |document|
+            post_process_result = post_processing_engine.process(document)
+            if post_process_result.is_include?
+              @sink.ingest(document)
+              incoming_ids << document['id']
+            end
           end
+        else
+          Utility::Logger.info('Yielding documents without basic rule filtering...')
 
-          periodically do
-            check_job
-            @job.update_metadata(@sink.ingestion_stats, @connector_instance.metadata)
+          yield_docs do |document|
+            @sink.ingest(document)
+            incoming_ids << document['id']
```

Comment:
Github doesn't support suggestions on deleted lines. I just feel this `if..else` contains duplicate code. I think we can have something like

```ruby
post_processing_engine = @connector_settings.filtering_rule_feature_enabled? ? Core::Filtering::PostProcessEngine.new(@job.filtering) : nil

yield_docs do |document|
   continue if post_processing_engine && !post_processing_engine.process(document).is_include?
   @sink.ingest(document)
   incoming_ids << document['id']
 end

```

Reply:
Looks a lot cleaner, adjusted it!